thetis/utility.py

```diff
@@ -306,11 +306,64 @@ def comp_tracer_mass_2d(eta, bath, scalar_func):
     :arg bath: bathymetry :class:`Function`
     :arg scalar_func: scalar :class:`Function` to integrate
     """
-
     val = assemble((eta+bath)*scalar_func*dx)
     return val
 
 
+def comp_tracer_total_mass_2d(var, tracer_name):
+    """
+    Computes total tracer mass in the 2D domain accounting for tracer leaving through
+    boundary conditions and source terms
+    :arg var: class:`DiagnosticCallback`; callback object used to input values and
+                                          record output values
+    :arg tracer_name :class:`string` of function name of interest
+    """
+
+    # read in necessary variables from solver object
+    eta = var.solver_obj.fields.elev_2d.copy(deepcopy=True)
+    vel = var.solver_obj.fields.uv_2d.copy(deepcopy=True)
+
+    scalar_func = var.solver_obj.fields[tracer_name].copy(deepcopy=True)
+
+    # calculate total depth
+    term = var.solver_obj.eq_tracer.terms['HorizontalAdvectionTerm']
+    H = term.get_total_depth(eta)
+
+    # normal
+    n = FacetNormal(var.solver_obj.mesh2d)
+
+    # calculate contribution from tracer leaving boundary
+    boundary_terms = 0
+
+    for bnd_marker in term.boundary_markers:
+        ds_bnd = ds(int(bnd_marker), degree=term.quad_degree)
+        a = -(n[0]*vel[0] + n[1]*vel[1])*H*scalar_func*ds_bnd
+
+        boundary_terms += assemble(a)
+
+    # record the initial scalar value in the domain
+    if var.initial_value is None:
+        var.initial_value = assemble(H*scalar_func*dx)
+        var.update_value = assemble(H*scalar_func*dx)
+    else:
+        # alter the initial value to record tracer transitioning through source term
+        # and boundary terms
+        var.update_value += var.solver_obj.options.timestep * boundary_terms
+
+        if var.solver_obj.options.tracer_source_2d is not None:
+            var.update_value += var.solver_obj.options.timestep * \
+                assemble(var.solver_obj.options.tracer_source_2d*H*dx)
+
+    # find the current scalar value in the domain
+    val = assemble(H*scalar_func*dx)
```

Comment:
If you wanted to create a separate function for computing the tracer mass, it would only contain this line. Which, in fact, is a generalization of `comp_tracer_mass_2d`.